src/odemis/acq/align/autofocus.py

```diff
@@ -365,8 +379,199 @@ def AutoFocus(detector, emt, focus, dfbkg=None, good_focus=None):
+    gratings, and for each detectors. The input slit should already be in a
+    good position (typically, almost closed), and a light source should be
+    active.
+    Note: it's currently tailored to the Andor Shamrock SR-193i. It's recommended
+    to put the detector on the "direct" output as first detector.
+    spectrograph (Actuator): should have grating and wavelength.
+    focuser (Actuator): should have a z axis
+    detectors (Detector or list of Detectors): all the detectors available on
+      the spectrometer. The first detector will be used to autofocus all the
+      gratings, and each other detector will be focused with the original
+      grating.
+    selector (Actuator or None): must have a rx axis with each position corresponding
+     to one of the detectors. If there is only one detector, selector can be None.
+    return (ProgressiveFuture -> dict((grating, detector)->focus position)): a progressive future
+      which will eventually return a map of grating/detector -> focus position.
+    """
+    if not isinstance(detectors, collections.Iterable):
+        detectors = [detectors]
+    if not detectors:
+        raise ValueError("At least one detector must be provided")
+
+    # Create ProgressiveFuture and update its state to RUNNING
+    est_start = time.time() + 0.1
+    detector = detectors[0]
+    if model.hasVA(detector, "exposureTime"):
+        et = detector.exposureTime.value
+    else:
+        # Completely random... but we are in a case where probably that's the last
+        # thing the caller will care about.
+        et = 1
+
+    # 1 time / grating + 1 time / extra detector
+    cnts = len(spectrograph.axes["grating"].choices) + (len(detectors) - 1)
+    f = model.ProgressiveFuture(start=est_start,
+                                end=est_start + cnts * estimateAutoFocusTime(et))
+    f.task_canceller = _CancelAutoFocusSpectrometer
+    # Extra info for the canceller
+    f._autofocus_state = RUNNING
+    f._autofocus_lock = threading.Lock()
+    f._subfuture = InstantaneousFuture()
+
+    # Run in separate thread
+    autofocus_thread = threading.Thread(target=executeTask,
+                                        name="Spectrometer Autofocus",
+                                        args=(f, _DoAutoFocusSpectrometer, f,
+                                              spectrograph, focuser, detectors, selector))
+
+    autofocus_thread.start()
+    return f
+
+
+def _moveSelectorToDetector(selector, detector):
+    """
+    Move the selector to have the given detector receive light
+    selector (Actuator): a rx axis with a position
+    detector (Component): the component to receive light
+    return (position): the new position of the selector
+    raise LookupError: if no position on the selector affects the detector
+    """
+    # TODO: handle every way of indicating affect position? -> move to odemis.util
+    choices = selector.axes["rx"].choices
```

Comment:
Isn't it a bit risky to assume that the axis is called "rx"? Maybe a more generic solution?